docs/performance/optimization.rst

```diff
@@ -0,0 +1,598 @@
+   FROM device_data
+   GROUP BY deviceid
+   LIMIT 1000 OFFSET 5000;
+
+We can do something like
+
+.. code:: sql
+
+   WITH devices AS (
+     SELECT deviceid
+     FROM devices
+     LIMIT 5 OFFSET 25
+   )
+   SELECT deviceid, AVG(field1)
+   FROM device_data
+   WHERE device_data.deviceid IN (SELECT devices.deviceid FROM devices)
+   GROUP BY deviceid;
+
+.. _staging-tables:
+
+Use staging tables for intermediate results if you are doing a lot of JOINs
+===========================================================================
+
+If you have many CTEs or VIEWs with a need to JOIN them, it can be benefical
+to query them individually, store intermediate results into dedicated tables, and
+then use these tables for JOINing.
+
+While there is a cost in writing to disk and reading data back, the whole operation
+can benefit from indexing and from giving the optimizer more straightforward
+execution plans, to enable it optimizing for better parallel execution using
+multiple cluster nodes.
+
+.. _group-schema-and-function-optimization:
+
+**********************************
+ Schema and Function Optimization
+**********************************
+
+This section focuses on schema design and function usage to streamline
+performance.
+
+.. _consider-generated-columns:
+
+Consider generated columns
+==========================
+
+If you frequently find yourself extracting information from fields and then
+using this extracted data on filters or aggregations, it can be good to consider
+doing this operation on ingestion with a `generated column`_, this way the value
+we need for filtering and aggregations can be indexed.
+
+See `Using regex comparisons and other features for inspection of logs`_ for an
+example.
+
+.. _udf-right-context:
+
+Be mindful of UDFs, leverage them in the right contexts, but only in the right contexts
+=======================================================================================
+
+When using user-defined functions (UDFs), two important details relevant for
+performance aspects need to be considered.
+
+#. Once values are processed by an UDF, the database engine will load results
+   into memory, and will not be able to leverage indexes on the underlying
+   fields any longer. In this spirit, please apply the relevant general
+   considerations about delaying formatting as much as possible.
+
+#. UDFs run on a JavaScript virtual machine on a single thread, so they can have
+   an impact on performance, because relevant operations can not be
+   parallelized.
+
+However, some operations may be more straightforward to do in JavaScript than
+SQL.
+
+.. _group-filter-expression-optimizations:
+
+This section discusses expressions that improve filter efficiency and handling
+of specific data Structures.
+
+************************************
+ Filter and Expression Optimization
+************************************
+
+.. _positive-filters:
+
+Prefer positive filter expressions to negative filter expressions
```

Comment:
Is this really called like that? Isn't this more popular as "Avoid expression negation"`

Reply:
Neither "negative filter expressions" nor "expression negation" seems particularly popular.
I changed the title to "Avoid expression negation in filters" now.
The underlying and more comprehensive technical concept is [sargability](https://en.wikipedia.org/wiki/Sargable), but I think we probably just want to keep it simple.